src/car.ts

```diff
@@ -79,86 +80,106 @@ export default class Car extends GameElement {
+            this.copy({
+              jump_timer: Math.max(0, this.jump_timer - 1)
             })) :
         (this.jump_state == "station" ?
-            cat_after_nitro_input.copy({
-              jump_timer: Math.max(0, cat_after_nitro_input.jump_timer - 1)
+            this.copy({
+              jump_timer: Math.max(0, this.jump_timer - 1)
             }) :
-            cat_after_nitro_input.copy({
+            this.copy({
               jump_state: "station",
-              jump_timer: Math.max(0, cat_after_nitro_input.jump_timer - 1)
+              jump_timer: Math.max(0, this.jump_timer - 1)
             }));
-    const car_after_all_input = car_after_jump_input;
 
-    const car_jumping_index = car_after_all_input.jump_timer;
-    const car_jumping = car_after_all_input.jump_timer > 0;
-    const car_flying = car_after_all_input.flying_state == "flying";
-    const nitro_active = car_after_all_flips.nitro_state == "active";
+    const car_jumping_index = after_jump_input.jump_timer;
+    const car_jumping = after_jump_input.jump_timer > 0;
+    const car_flying = after_jump_input.flying_state == "flying";
 
-    const nitro_vector = car_after_all_flips.nitro.normalize()
-        .multiply(nitro_active ? 1 : 0).rotate(this.angle).reverse().multiply(20);
-    const jump_vector = car_after_all_flips.jumper.normalize()
+    const jump_vector = this.jumper.normalize()
         .multiply(car_jumping ? 1 : 0).rotate(this.angle).reverse().multiply(20);
-    const angular_force = (Constants.key_pressed.get("left") * -1 +
-        Constants.key_pressed.get("right")) * (car_flying ? 4 : 0);
 
-    const advanced_car: Car = car_after_all_input.copy({
-        velocity: car_after_all_input.velocity.add_vector(nitro_vector.multiply(time_unit * 1.0 / 1000))
-            .add_vector(jump_vector.multiply(1.0 / Math.pow(2, Constants.jump_timer_duration - car_jumping_index + 1))),
-        angular_velocity: car_after_all_input.angular_velocity +
-            (angular_force * time_unit * 1.0 / 1000)
+    return after_jump_input.copy({
+        velocity: after_jump_input.velocity.add_vector(jump_vector.multiply(1.0 / Math.pow(2, Constants.jump_timer_duration - car_jumping_index + 1))),
     });
+  }
 
-    const advanced_car_original_position = advanced_car.copy<Car>({
-      position: this.position,
-      angle: this.angle
+  private _apply_rotation_input(time_unit: number): Car {
+    const angular_force = (Constants.key_pressed.get("left") * -1 +
+        Constants.key_pressed.get("right"));
+    const rotation_strength = 4;
+
+    return this.copy({
+        angular_velocity: this.angular_velocity + (rotation_strength * angular_force * time_unit * 1.0 / 1000)
     });
-    return advanced_car_original_position.collideAll(
-        advanced_car.position.subtract(this.position), advanced_car.angle - this.angle, game_set);
+  }
+
+  public updated(time_unit: number): GameElement {
+    const start_state = this;
+
+    const pipeline = Immutable.List([
```

Comment:
Didn't get your suggestion? It's the problem of calling successive methods on immutable objects where the result of a method needs to be passed as input to the successor method and how that chaining happens.

A better representation for this maybe is to wrap this logic in a Pipeline class so that the usage goes like this:

const pipeline = new Pipeline([
      { method: this._apply_flip_x_input, parameters: [] },
      { method: this._apply_flip_y_input, parameters: [] },
      { method: this._apply_nitro_input, parameters: [time_unit] },
      { method: this._apply_jump_input, parameters: [time_unit] },
      { method: this._apply_rotation_input, parameters: [time_unit] },
      { method: this._updated_physics, parameters: [time_unit] }
    ]);
const final_state = pipeline.execute(initial_state);